### PR TITLE
Added internal sol for icy perimeter silver

### PR DESCRIPTION
**Pull request by TanishT**
I noticed that there was no internal sol, so I thought it would be nice to add one that if different than the editorial.

### CODE REVIEW

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (bqi343) [inline solutions/USACO jan19-icy-perimeter.mdx:24]: Prefer `using ll = long long` and `const int MAXN = 1005, MAXM = 1005;` as noted here:

https://usaco.guide/general/generic/#macros

(also, is it necessary to have separate variables for `MAXN` and `MAXM` when the grid is square?)

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Comment by bqi343** [comment]
You also need to do step 4 of https://usaco.guide/general/contributing#adding-a-solution

**Comment by dolphingarlic** [comment]
Isn't there already a PR for this problem?

**Comment by TanishT** [comment]
> Isn't there already a PR for this problem?

Uhh I didn't see this when I was submitting the PR. Should I just close this one?

**Comment by maggie-j-liu** [comment]
There's another PR for this problem (#932), and the official solution is also C++...

**Comment by bqi343** [comment]
Well, it's different in the sense that this one actually uses recursion. I'm fine with adding it.

**Review by bqi343 — APPROVED** [review]

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [inline solutions/USACO jan19-icy-perimeter.mdx:80]: I don't think this is right... consider an input like
```
6
###.#.
###.#.
###.#.
....#.
....#.
....#.
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (jessechoe10) [inline solutions/USACO jan19-icy-perimeter.mdx:28]: `ll curA = 0; // Keeps track of the current area of the current blob of ice cream'.

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (jessechoe10) [inline solutions/USACO jan19-icy-perimeter.mdx:29]: `ll curP= 0; // Keeps track of the current perimeter of the blob of ice cream`.

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (jessechoe10) [inline solutions/USACO jan19-icy-perimeter.mdx:30]: Change to `ll ans1 = 0` - optional...

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (jessechoe10) [inline solutions/USACO jan19-icy-perimeter.mdx:31]: Change to `ll ans2 = 0` - optional...

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (jessechoe10) [inline solutions/USACO jan19-icy-perimeter.mdx:64]: change to 
`freopen("perimeter.in", "r", stdin);`
and
`freopen("perimeter.out", "w", stdout);`

Comment (jessechoe10) [reply solutions/USACO jan19-icy-perimeter.mdx:64]: the setIO() function is kinda weird so you should make it concise like this :stuck_out_tongue:

**Review by jessechoe10 — CHANGES_REQUESTED** [review]

**Comment by jessechoe10** [comment]
Perhaps this code might be better idk (totally up to everyone else):

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;
using vb = vector<bool>;
using pl = pair<ll, ll>;
using vpl = vector<pl>;
using vc = vector<char>;
using vs = vector<string>;
using str = string;

#define FastIO ios_base::sync_with_stdio(0); cin.tie(0)
#define pb push_back
#define mp make_pair
#define all(x) begin(x), end(x)
#define sor(x) sort(all(x))
#define ins insert
#define f first
#define s second
#define sz(x) (ll)(x).size()
#define len(x) (ll)(x).length()
#define FOR(i, a, b) for(ll i=a; i<b; ++i)
#define ROF(i, b, a) for(ll i=(b)-1; i>=a; --i)
#define EACH(a, x) for(auto& a: x)

void setIO(string fileName = ""){
	FastIO;
	if(len(fileName)){
		freopen((fileName+".in").c_str(), "r", stdin);
		freopen((fileName+".out").c_str(), "w", stdout);
	}
}

char icecream[1001][1001];
bool visited[1001][1001];

ll n, perimeter, area;

bool cmp(pl a, pl b){
	if(a.f == b.f) return a.s > b.s;
	return a.f < b.f;
}

void DFS(ll i, ll j){
	if(i >= n || j >= n || i < 0 || j < 0 || visited[i][j] || icecream[i][j] == '.') return;
	visited[i][j] = true;
	if(icecream[i][j] == '#') area++;
	if(icecream[i-1][j] != '#') perimeter++;
	if(icecream[i+1][j] != '#') perimeter++;
	if(icecream[i][j-1] != '#') perimeter++;
	if(icecream[i][j-1] != '#') perimeter++;
	DFS(i-1, j); DFS(i+1, j); DFS(i, j-1); DFS(i, j+1);
}

int main(){
	setIO("perimeter");
	cin >> n;
	FOR(i, 0, n){
		FOR(j, 0, n){
			cin >> icecream[i][j];
		}
	}
	vpl properties;
	FOR(i, 0, n){
		FOR(j, 0, n){
			perimeter = 0, area = 0;
			DFS(i, j);
			if(perimeter && area) properties.pb(mp(area, perimeter));
		}
	}
	sort(all(properties), cmp);
	cout << properties[sz(properties)-1].f << " " << properties[sz(properties)-1].s << endl;
}
```

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [reply solutions/USACO jan19-icy-perimeter.mdx:30]: I think `area` and `perimeter` are more descriptive... but yeah, no need to use `INT_MIN`

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [inline solutions/USACO jan19-icy-perimeter.mdx:30]: this comment is in the wrong place, it should be for `currA`

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (jessechoe10) [reply solutions/USACO jan19-icy-perimeter.mdx:28]: pls address this

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [reply solutions/USACO jan19-icy-perimeter.mdx:79]: This hasn't been fixed yet...

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [inline solutions/USACO jan19-icy-perimeter.mdx:29]:
```suggestion
ll currA = 0; // keeps track of the current area of the current blob of ice cream
ll currP = 0; // keeps track of the current perimeter of the blob of ice cream
```

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [inline solutions/USACO jan19-icy-perimeter.mdx:17]:
```suggestion
// created by Tanish Tyagi
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Review by jessechoe10 — APPROVED** [review]

**Comment by jessechoe10** [comment]
@frodakcin I think we should merge this unless you have any objections (two approvals)...

**Comment by maggie-j-liu** [comment]
I think there's a problem with the solution... consider an input like this:
```
6
###.#.
###.#.
###.#.
....#.
....#.
....#.
```
Although it's very minor and can be easily fixed

**Comment by jessechoe10** [comment]
> I think there's a problem with the solution... consider an input like this:
> 
> ```
> 6
> ###.#.
> ###.#.
> ###.#.
> ....#.
> ....#.
> ....#.
> ```
> 
> Although it's very minor and can be easily fixed

Perhaps my solution may be better:

```cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using vl = vector<ll>;
using vb = vector<bool>;
using pl = pair<ll, ll>;
using vpl = vector<pl>;
using vc = vector<char>;
using vs = vector<string>;
using str = string;

#define FastIO ios_base::sync_with_stdio(0); cin.tie(0)
#define pb push_back
#define mp make_pair
#define all(x) begin(x), end(x)
#define sor(x) sort(all(x))
#define ins insert
#define f first
#define s second
#define sz(x) (ll)(x).size()
#define len(x) (ll)(x).length()
#define FOR(i, a, b) for(ll i=a; i<b; ++i)
#define ROF(i, b, a) for(ll i=(b)-1; i>=a; --i)
#define EACH(a, x) for(auto& a: x)

void setIO(string fileName = ""){
	FastIO;
	if(len(fileName)){
		freopen((fileName+".in").c_str(), "r", stdin);
		freopen((fileName+".out").c_str(), "w", stdout);
	}
}

char icecream[1001][1001];
bool visited[1001][1001];

ll n, perimeter, area;

bool cmp(pl a, pl b){
	if(a.f == b.f) return a.s > b.s;
	return a.f < b.f;
}

void DFS(ll i, ll j){
	if(i >= n || j >= n || i < 0 || j < 0 || visited[i][j] || icecream[i][j] == '.') return;
	visited[i][j] = true;
	if(icecream[i][j] == '#') area++;
	if(icecream[i-1][j] != '#') perimeter++;
	if(icecream[i+1][j] != '#') perimeter++;
	if(icecream[i][j-1] != '#') perimeter++;
	if(icecream[i][j-1] != '#') perimeter++;
	DFS(i-1, j); DFS(i+1, j); DFS(i, j-1); DFS(i, j+1);
}

int main(){
	setIO("perimeter");
	cin >> n;
	FOR(i, 0, n){
		FOR(j, 0, n){
			cin >> icecream[i][j];
		}
	}
	vpl properties;
	FOR(i, 0, n){
		FOR(j, 0, n){
			perimeter = 0, area = 0;
			DFS(i, j);
			if(perimeter && area) properties.pb(mp(area, perimeter));
		}
	}
	sort(all(properties), cmp);
	cout << properties[sz(properties)-1].f << " " << properties[sz(properties)-1].s << endl;
}
```

**Comment by jessechoe10** [comment]
> > I think there's a problem with the solution... consider an input like this:
> > ```
> > 6
> > ###.#.
> > ###.#.
> > ###.#.
> > ....#.
> > ....#.
> > ....#.
> > ```
> > 
> > 
> > Although it's very minor and can be easily fixed
> 
> Perhaps my solution may be better:
> 
> ```c++
> #include <bits/stdc++.h>
> 
> using namespace std;
> using ll = long long;
> using vl = vector<ll>;
> using vb = vector<bool>;
> using pl = pair<ll, ll>;
> using vpl = vector<pl>;
> using vc = vector<char>;
> using vs = vector<string>;
> using str = string;
> 
> #define FastIO ios_base::sync_with_stdio(0); cin.tie(0)
> #define pb push_back
> #define mp make_pair
> #define all(x) begin(x), end(x)
> #define sor(x) sort(all(x))
> #define ins insert
> #define f first
> #define s second
> #define sz(x) (ll)(x).size()
> #define len(x) (ll)(x).length()
> #define FOR(i, a, b) for(ll i=a; i<b; ++i)
> #define ROF(i, b, a) for(ll i=(b)-1; i>=a; --i)
> #define EACH(a, x) for(auto& a: x)
> 
> void setIO(string fileName = ""){
> 	FastIO;
> 	if(len(fileName)){
> 		freopen((fileName+".in").c_str(), "r", stdin);
> 		freopen((fileName+".out").c_str(), "w", stdout);
> 	}
> }
> 
> char icecream[1001][1001];
> bool visited[1001][1001];
> 
> ll n, perimeter, area;
> 
> bool cmp(pl a, pl b){
> 	if(a.f == b.f) return a.s > b.s;
> 	return a.f < b.f;
> }
> 
> void DFS(ll i, ll j){
> 	if(i >= n || j >= n || i < 0 || j < 0 || visited[i][j] || icecream[i][j] == '.') return;
> 	visited[i][j] = true;
> 	if(icecream[i][j] == '#') area++;
> 	if(icecream[i-1][j] != '#') perimeter++;
> 	if(icecream[i+1][j] != '#') perimeter++;
> 	if(icecream[i][j-1] != '#') perimeter++;
> 	if(icecream[i][j-1] != '#') perimeter++;
> 	DFS(i-1, j); DFS(i+1, j); DFS(i, j-1); DFS(i, j+1);
> }
> 
> int main(){
> 	setIO("perimeter");
> 	cin >> n;
> 	FOR(i, 0, n){
> 		FOR(j, 0, n){
> 			cin >> icecream[i][j];
> 		}
> 	}
> 	vpl properties;
> 	FOR(i, 0, n){
> 		FOR(j, 0, n){
> 			perimeter = 0, area = 0;
> 			DFS(i, j);
> 			if(perimeter && area) properties.pb(mp(area, perimeter));
> 		}
> 	}
> 	sort(all(properties), cmp);
> 	cout << properties[sz(properties)-1].f << " " << properties[sz(properties)-1].s << endl;
> }
> ```

I can detemplify this if you're gonna add this....

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (TanishT) [reply solutions/USACO jan19-icy-perimeter.mdx]: just fixed it, thanks for pointing it out!

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [inline solutions/USACO jan19-icy-perimeter.mdx:29]:
```suggestion
ll currA = 0; // keeps track of the current area of the current blob of ice cream
ll currP = 0; // keeps track of the current perimeter of the blob of ice cream
```
this was somehow reverted?

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [reply solutions/USACO jan19-icy-perimeter.mdx:23]: please address this

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/USACO jan19-icy-perimeter.mdx
Comment (maggie-j-liu) [reply solutions/USACO jan19-icy-perimeter.mdx:23]: actually, I think you did but the changes were reverted?

**Comment by maggie-j-liu** [comment]
I think some of the requested changes were reverted by a later commit, could you check and make sure you applied all the requested changes?

**Review by maggie-j-liu — APPROVED** [review]

**Comment by jessechoe10** [comment]
@thecodingwizard can we merge this solution despite the required authorization. if so, i think we can merge rn cuz we have three approvals...

**Pull request merged**
